### PR TITLE
Combine authorization for Okta and Server2server

### LINKED ISSUE #5290: Consolidate Group-based and Scope-based Claims
### Description/Use Case

This is tech debt left over from #5132, which enable both server2server and otka auth across a number of azure functions.

The main problem is that Okta sends "organization" (Okta Group) claims, and server2server auth sends "scope" claims.  This ticket is to do behind-the-scenes refactoring so that, internally, both styles are tracked as "scope" claims.

### Risks/Impacts/Considerations

- We need to implement both styles of auth for many other functions, in particular, the settings functions.  Before we do that, we should do this ticket.
- Note that the *.*.primeadmin scope doesn't make much sense for server2server auth, because, currently, all server2server authenticated scopes are associated with specific orgs, in the settings table.

### Acceptance Criteria

- The work described below is done.

#### Current work:

- [x] write a design markdown doc for what the new AuthenticatedClaims will look like.  Include use cases for both current settings queries and future delivery etc queries, public use, etc.   Basically, answer this question:  how do we make the "DH" okta style Groups look like the server2server "scopes", and how do we represent that in code?
- [x] fully consolidate uses of AuthenticatedClaims so that both Server2Server Auth and Okta Auth return that object, and the fields are consolidated between the two
- [x] Enrich the functionality of Scope.kt so it can be used for both Okta and Server2server auth
- [x] For some reason I wrote unit tests for Scope.kt, and put them in other places.  Consolidate and write further unit tests.
- [x] Tie scope to Claims.
- [x] Rename TokenAuthentication to Server2ServerAuth
- [x] somehow consolidate AuthenticatedClaims, AuthenticationStrategy, Scope
- [x] write a smoke test that tests server2server cannot upgrade its auth to primeAdmin.
- [X] Can I just get rid of AuthenticatedClaims.isSenderOrgClaim???  
- [X] The unit tests for Scope.kt are in wrong classes.   Create a new ScopeTests.kt, consolidate tests.
- [X] The two constructors in AuthenticatedClaims.kt are now redundant.  Replace with one.
- [x] write a function that tests for general submitting/delivery level authorization, and works for both Okta and Server2server.
- [x] get rid of TokenAuthentication.checkAccessToken - it is not used by real code now that authz/authw have been split.

### Other cleanup work to scopes and okta group based permissions, not in current work.

Clean up this list and Consolidate with #1946.

- [ ] Get rid of the rather ugly`isOktaAuth` (and related `authentication-type: okta` header!), and simply check for whether the user sent `organization` Okta claims or `scope` Server2Server claims.  A bit of a hack, but worth it to avoid the annoying header. 
- [ ] Change Settings and other functions to use this new "both server2server and okta" auth.  Delete OktaAuthentication.checkAccess and OktaAuthentication.authorizeByMembership, and replace with the more general functions in AuthenticatedScope.kt.
- [ ] Enable the server2server keys section to handle multiple scopes per key.   Enable authorization to handle a scope with multiple keys.  Is it time to have a completely separate keys/scope table?
- [ ] Very minor: Make all Methods in TokenAuthentication be static (eg, companion object).  There's no reason to be creating TokenAuthentication objects all over the place.
- [ ] Need to write a unit test for non-null client value in OktaAuthentication.validate(...., client)

Other Tech debt
- [ ] Do a general pass through and clean up our security related logging, make it consistent, and maybe consolidate it with our http response error handling mechanism.
- [ ] The oktaauth test does not work in Staging.   Two issues:  the localauth tests are only meant to test local auth, so pointless to run them on staging.  Also, oktaauth uses local auth to create the test reports it uses, which doesn't work on Staging.  Both of these are pretty straightforward to fix.

### PR DESCRIPTION
**Pull request by jimduff-usds**
This PR fixes the problem that Okta and Server2server auth had two different ways of tracking claims, which required two different ways of doing Authorization, which made it very hard to implement both types of Auth in a single Azure function.

Okta sends "organization" (Okta Group) claims, and server2server auth sends "scope" claims, per our "FHIR Auth standard". This PR is behind-the-scenes refactoring so that, internally, both styles are tracked as "scope" claims.

Currently only the Waters submission, the Delivery, and the Submission API, and ValidateFunction endpoints implement both styles of Auth.   This PR is prep work for implementing both styles of auth for many other functions, in particular, the settings functions. 

### For more info

Background information can be found in the 0016-authorization.md proposal doc, in this PR. (https://github.com/CDCgov/prime-reportstream/blob/jim/settings-auth/prime-router/docs/proposals/0016-authorization.md)
See #5290 for detailed work done.

### Test Steps:
1. In addition to running regular smoke tests on this branch, be sure to run the okta auth test, not currently part of the regular smokes:

```
./prime test    (Run the regular smoke test)
./prime login --staging    (Be sure to login with a primeadmin user)
./prime test --run oktauth
```

2. Test the UI.

## Changes
- A very comprehensive list of changes can be found in #5290.

### Testing
- [X] Tested locally?
- [X] Ran `./prime test` or `./gradlew testSmoke` against local Docker ReportStream container?
- [ ] (For Changes to /frontend-react/...) Ran `npm run lint:write`? 
- [X] Added tests?

### Process
- [N] Are there licensing issues with any new dependencies introduced?
- [Y] Includes a summary of what a code reviewer should test/verify?
- [N] Updated the release notes?
- [N/A] Database changes are submitted as a separate PR?
- [N/A] DevOps team has been notified if PR requires ops support?

## Linked Issues
- Fixes #5290 

## To Be Done
- Follow-on work is in #1946, and tickets linked from there.

## Specific Security-related subjects a reviewer should pay specific attention to
- Does this PR include changes in authentication and/or authorization of existing endpoints? 
**Answer: YES**

If you answered '_yes_' to any of the questions above, conduct a detailed Review that addresses at least:

- What are the potential security threats and mitigations? Please list the _STRIDE_ threats and how they are mitigated

**Answer:**

Potential threats are Spoofing and Elevation of Privilege.

Mitigation is a comprehensive set of unit tests, and two large smoke tests:
./prime test --run server2serverauth,oktaauth

Tests for elevation of privilege have been added to the smoke tests.

- Have you ensured logging does not contain sensitive data?  **YES**
- Have you received any additional approvals needed for this change?  **YES**

### CODE REVIEW

File: prime-router/src/main/kotlin/cli/tests/WatersAuthTests.kt [hunk 1]
```diff
@@ -708,64 +709,108 @@ class Server2ServerAuthTests : CoolTest() {
...
+        // Test various functionality using the general <orgname>.*.admin role.
+        val scope2 = "$org2.*.admin"
         // Submit this new scope and public key to the Settings store, associated with this Sender.
         sender2 = saveServer2ServerKey(sender2, end2EndExampleRSAPublicKeyStr, kid, scope2)
-        sender2 = saveServer2ServerKey(sender2, end2EndExampleRSAPublicKeyStr, kid, uploadReportScope2)
 
         try {
             val myFakeReportFile = OktaAuthTests.createFakeReport(sender1)
 
-            // Now request 5-minute tokens for the first org
```
Comment (jimduff-usds) [inline prime-router/src/main/kotlin/cli/tests/WatersAuthTests.kt:732]: `uploadReportScope1` is a nice example of a very narrowly defined scope - the ability to upload to this specific receiver with this specific org.

File: prime-router/src/main/kotlin/tokens/AuthenticatedClaims.kt [hunk 2]
```diff
@@ -1,135 +1,227 @@
...
     /** Name of user as extracted from the subject claim.  Usually an email address */
     val userName: String
+
     /** Does this user have the prime administrator claim? */
     val isPrimeAdmin: Boolean
-    /** Name of the organization found in these claims */
-    val organizationNameClaim: String?
-    /** Is this a Sender claim, of the form DHSender_orgname  ? */
-    val isSenderOrgClaim: Boolean
+
```
Comment (jimduff-usds) [inline prime-router/src/main/kotlin/tokens/AuthenticatedClaims.kt:35]: This `scopes` is the main new attribute added to AuthenticatedClaims.

File: prime-router/src/main/kotlin/tokens/AuthenticatedClaims.kt [hunk 3]
```diff
@@ -1,135 +1,227 @@
...
 
     /**
-     * This constructor assumes certain claims that are found in Okta. Only use this for Okta claims.
+     * [_jwtClaims]  The raw list of claims
+     * [isOktaAuth] true if the claims came from Okta (eg, groups like "DHmd-phd").  false if the claims came from
+     * server2server auth (eg, scopes like md-phd.default.report)
+     *
+     * Each value passed must have been already authenticated.  This is for claims, *not* for required authorizations.
      */
-    constructor(_jwtClaims: Map<String, Any>) {
```
Comment (jimduff-usds) [inline prime-router/src/main/kotlin/tokens/AuthenticatedClaims.kt:32]: In a future PR, we can get rid of this rather ugly`isOktaAuth` (and related `authentication-type: okta` header!), and simply check for whether the user sent `organization` Okta claims or `scope` Server2Server claims.  A bit of a hack, but worth it to avoid the annoying header.

File: prime-router/src/main/kotlin/tokens/AuthenticatedClaims.kt [hunk 4]
```diff
@@ -1,135 +1,227 @@
...
      */
-    private fun extractOrganizationNameFromOktaMembership(memberships: Collection<String>): Pair<String?, Boolean>? {
-        if (memberships.isEmpty()) return null
-        // examples:   DHSender_ignore, DHPrimeAdmins, DHSender_all-in-one-health-ca, DHaz_phd, DHignore
-        // should return, resp.:  (ignore, true), <nothing>, (all-in-one-health-ca, true, (az_phd, false, (ignore, true)
-        memberships.forEach {
-            if (it == oktaSystemAdminGroup) return@forEach // skip
-            if (it.startsWith(oktaSenderGroupPrefix)) return Pair(it.removePrefix(oktaSenderGroupPrefix), true)
-            if (it.startsWith(oktaGroupPrefix)) return Pair(it.removePrefix(oktaGroupPrefix), false)
+    fun authorizedForSubmission(
```
Comment (jimduff-usds) [inline prime-router/src/main/kotlin/tokens/AuthenticatedClaims.kt:90]: This function "knows" all the scopes that are allowed to do Submission related activities.

File: prime-router/src/main/kotlin/tokens/AuthenticatedClaims.kt [hunk 5]
```diff
@@ -1,135 +1,227 @@
...
-            if (it == oktaSystemAdminGroup) return true
-        }
-        return false
+    fun authorized(requiredScopes: Set<String>): Boolean {
+        return Scope.authorized(this.scopes, requiredScopes)
     }
 
-    companion object {
+    companion object : Logging {
         /**
```
Comment (jimduff-usds) [inline prime-router/src/main/kotlin/tokens/AuthenticatedClaims.kt:137]: This big swath of green code here is not really new, its copied over from AuthenticationStrategy.kt, which has been deleted.

File: prime-router/src/main/kotlin/tokens/OktaAuthentication.kt [hunk 6]
```diff
@@ -99,9 +73,72 @@ class OktaAuthentication(private val minimumLevel: PrincipalLevel = PrincipalLev
             // Perform authentication.  Throws exception if authentication fails.
             return jwtVerifier.decode(accessToken)
         }
+
+        /**
```
Comment (jimduff-usds) [inline prime-router/src/main/kotlin/tokens/OktaAuthentication.kt:77]: This function is still widely used, for example by Settings API and LookupTable API and others.

However, once this PR is in, we can begin to migrate those over to the dual-authentication and authorization functions found in AuthenticatedClaims.kt, rather than this Okta-specific code.

File: prime-router/src/main/kotlin/tokens/Scope.kt [hunk 7]
```diff
@@ -16,26 +15,71 @@ class Scope {
...
          * Scope that represents prime admin access.
+         * Pattern is <organization>.<receiver-or-sender>.<DetailedScope>
+         * The first value must always be an organization name, or wildcard
+         * The idea is to leave the last field loose; it could be a role or a permission
+         * Examples:
+         * md-phd.*.user -- user level access
+         * md-phd.default.report -- able to submit reports only
+         * md-phd.*.admin
+         * *.*.primeadmin
          */
```
Comment (jimduff-usds) [inline prime-router/src/main/kotlin/tokens/Scope.kt:26]: Changed per the 0016-authorization.md proposal.

File: prime-router/src/main/kotlin/tokens/Scope.kt [hunk 8]
```diff
@@ -16,26 +15,71 @@ class Scope {
...
-        val primeAdminScope = "primeadmin.all.all"
+        val primeAdminScope = "*.*.primeadmin"
 
-        enum class DetailedScope {
-            report, // ability to submit a report
-            all, // all abilities
```
Comment (jimduff-usds) [inline prime-router/src/main/kotlin/tokens/Scope.kt:24]: A Scope is just a string, of the form `x.y.<DetailScope>`

This is purposely very broad, to allow for future use cases no-one has dreamed of yet.

File: prime-router/src/main/kotlin/tokens/Scope.kt [hunk 9]
```diff
@@ -72,19 +114,93 @@ class Scope {
             val splits = scope.split(".")
             val orgName = splits[0]
             val senderOrReceiver = splits[1]
-            val detailedScope = DetailedScope.valueOf(splits[2])
+            val detailedScope: DetailedScope = DetailedScope.toDetailedScope(splits[2])!! // checked in isValidScope
             return Triple(orgName, senderOrReceiver, detailedScope)
         }
 
         /**
```
Comment (jimduff-usds) [inline prime-router/src/main/kotlin/tokens/Scope.kt:121]: All calls to authorize anything in reportstream should end up calling `authorizedScopes` here, which is intentionally very dumb - it makes no attempt know anything about the meaning of the scope strings.

If there is any intersection between the authenticated scopes the user claims, and the scopes required to use a resource, then the user is authorized.

File: prime-router/src/main/kotlin/tokens/Scope.kt [hunk 10]
```diff
@@ -72,19 +114,93 @@ class Scope {
...
+                        " required scope is: (${requiredScopeList.joinToString(",")})"
+                )
+                false
+            } else {
+                logger.info("User authorized for scopes (${intersection.joinToString(",")})")
+                true
+            }
+        }
+
+        /**
```
Comment (jimduff-usds) [inline prime-router/src/main/kotlin/tokens/Scope.kt:150]: This function is used by the TokenFunction code to lookup scopes in the Settings, which are associated with public keys.

File: prime-router/src/main/kotlin/tokens/Scope.kt [hunk 11]
```diff
@@ -72,19 +114,93 @@ class Scope {
...
          */
         fun scopeListContainsScope(scopeList: String, requiredScope: String): Boolean {
             if (requiredScope.isBlank() || requiredScope.isEmpty()) return false
             // A scope is a set of strings separated by single spaces
-            val scopesTrial: List<String> = scopeList.split(" ")
+            val scopesTrial: Set<String> = scopeList.split(" ").toSet()
             return scopesTrial.contains(requiredScope)
         }
+
+        /**
```
Comment (jimduff-usds) [inline prime-router/src/main/kotlin/tokens/Scope.kt:164]: This is the main function that maps from the world of Okta Groups to scopes.

File: prime-router/src/main/kotlin/tokens/Server2ServerAuthentication.kt [hunk 12]
```diff
@@ -148,15 +122,11 @@ class TokenAuthentication : Logging {
      *
      * @return the authenticated JWT Claims set if authentication was both successful.  Otherwise returns null.
      */
```
Comment (jimduff-usds) [inline prime-router/src/main/kotlin/tokens/Server2ServerAuthentication.kt:124]: This is a key change.  The old code was returning a server2server specific claims obj; the new code is returning our AuthenticatedClaims obj, which completely hides whether the authentication was done via Okta or Server2server.

File: prime-router/src/test/kotlin/tokens/Server2ServerAuthenticationTests.kt [hunk 13]
```diff
@@ -304,48 +312,38 @@ class TokenAuthenticationTests {
...
         val rslookup1 = GetTestSecret()
-        val accessToken = tokenAuthentication.createAccessToken("MyScope", rslookup1)
+        val accessToken = server2ServerAuthentication.createAccessToken("MyScope", rslookup1)
         val rslookup2 = GetTestSecret() // new/different secret
-        val claims = tokenAuthentication.checkAccessToken(accessToken.accessToken, "MyScope", rslookup2)
+        val claims = server2ServerAuthentication.authenticate(accessToken.accessToken, rslookup2)
         // if claims is non-null then the sender's accessToken is valid.
         assertNull(claims)
     }
-
```
Comment (jimduff-usds) [inline prime-router/src/test/kotlin/tokens/Server2ServerAuthenticationTests.kt:340]: This was moved to ScopeTests.kt.

File: prime-router/src/main/kotlin/tokens/OktaAuthentication.kt [hunk 14]
```diff
@@ -18,9 +16,9 @@ const val oktaGroupPrefix = "DH"
 const val oktaSenderGroupPrefix = "DHSender_"
 const val oktaAdminGroupSuffix = "Admins"
 const val oktaSystemAdminGroup = "DHPrimeAdmins"
-const val oktaSubjectClaim = "sub"
 const val oktaMembershipClaim = "organization"
 const val envVariableForOktaBaseUrl = "OKTA_baseUrl"
+const val subjectClaim = "sub"
```
Comment (clediggins-usds) [inline prime-router/src/main/kotlin/tokens/OktaAuthentication.kt:21]: Constants that are common between AuthenticatedClaims and OktaAuthentication could be pulled out into a AuthConstants.kt file

File: prime-router/src/main/kotlin/tokens/AuthenticatedClaims.kt [hunk 15]
```diff
@@ -1,135 +1,227 @@
...
+        request: HttpRequestMessage<String?>
+    ): Boolean {
+        return authorizedForSubmission(requiredSender.organizationName, requiredSender.name, request)
     }
 
     /**
-     * Derive useful info from jwtClaims on [memberships].
-     * @return the *first* well-formed ReportStream organizationName found in [memberships].
-     * (So this won't work if user has many organization claims.)
+     * Determine if these claims authorize access to the /validate and /waters endpoints for
```
Comment (clediggins-usds) [inline prime-router/src/main/kotlin/tokens/AuthenticatedClaims.kt:84]: Update comment to include /report endpoint as well, since that is a submission endpoint

File: prime-router/src/main/kotlin/tokens/Scope.kt [hunk 16]
```diff
@@ -16,26 +15,71 @@ class Scope {
...
+         * The first value must always be an organization name, or wildcard
+         * The idea is to leave the last field loose; it could be a role or a permission
+         * Examples:
+         * md-phd.*.user -- user level access
+         * md-phd.default.report -- able to submit reports only
+         * md-phd.*.admin
+         * *.*.primeadmin
          */
-        val primeAdminScope = "primeadmin.all.all"
+        val primeAdminScope = "*.*.primeadmin"
```
Comment (clediggins-usds) [inline prime-router/src/main/kotlin/tokens/Scope.kt:27]: Could be moved to an authConstants file with other auth constant values, so all hard coded constants for auth are in one location.

File: prime-router/src/main/kotlin/tokens/OktaAuthentication.kt [hunk 17]
```diff
@@ -99,9 +73,72 @@ class OktaAuthentication(private val minimumLevel: PrincipalLevel = PrincipalLev
...
+            logger.warn(
+                "User ${claims.userName}" +
+                    " memberships $membershipsFromOkta did NOT match requested $requiredMemberships"
+            )
+            return false
+        }
     }
 
     /**
+     * This is DEPRECATED.  Please stop using this, as it only does Okta auth, and does not allow for
```
Comment (clediggins-usds) [inline prime-router/src/main/kotlin/tokens/OktaAuthentication.kt:139]: Were follow-up tickets created for migrating/removing the deprecated logic we should no longer be using?

File: prime-router/src/main/kotlin/tokens/Scope.kt [hunk 18]
```diff
@@ -16,26 +15,71 @@ class Scope {
...
+
+        /**
+         * Return true if scope is the PrimeAdmin scope
+         */
+        fun isPrimeAdmin(scope: String?): Boolean {
+            return scope == primeAdminScope
         }
 
+        /**
+         * @return true if this [scope] syntax is correct.  Otherwise false.
```
Comment (clediggins-usds) [inline prime-router/src/main/kotlin/tokens/Scope.kt:58]: Is a wildcard in the third position valid syntax?  IE *.*.* ?

File: prime-router/src/main/kotlin/tokens/Scope.kt [hunk 19]
```diff
@@ -72,19 +114,93 @@ class Scope {
...
         }
 
         /**
+         * Finds the set of scopes that the user is authorized to have.
+         * Returns the matches between the set of scopes the user claims ([userScopeList])
+         * and the set of scopes acceptable to access a resource ([requiredScopeList])
+         *
+         * @return the (possibly empty) intersection of the two sets.  Empty Set == not authorized for anything.
+         */
+        fun authorizedScopes(userScopes: Set<String>, requiredScopes: Set<String>): Set<String> {
```
Comment (clediggins-usds) [inline prime-router/src/main/kotlin/tokens/Scope.kt:128]: should this be private / internal to prevent future accidental abuse?

File: prime-router/src/main/kotlin/tokens/Scope.kt [hunk 20]
```diff
@@ -72,19 +114,93 @@ class Scope {
...
+            return userScopes.filter { it.isNotBlank() }.intersect(requiredScopes.filter { it.isNotBlank() }.toSet())
+        }
+
+        /**
+         * @return true if there is at least one match between the set of scopes the user claims ([userScopeList])
+         * and the set of scopes acceptable to access a resource ([requiredScopeList])
+         */
+        fun authorized(userScopeList: Set<String>, requiredScopeList: Set<String>): Boolean {
```
Comment (clediggins-usds) [inline prime-router/src/main/kotlin/tokens/Scope.kt:136]: internal, to prevent accidental future abuse?

**Review by clediggins-usds — APPROVED** [review]

**Comment by TomNUSDS** [comment]
image

That's a lot of code smells?

File: prime-router/src/main/kotlin/tokens/OktaAuthentication.kt (hunk 14 above)
Comment (jimduff-usds) [reply prime-router/src/main/kotlin/tokens/OktaAuthentication.kt:21]: So moved.

File: prime-router/src/main/kotlin/tokens/AuthenticatedClaims.kt (hunk 15 above)
Comment (jimduff-usds) [reply prime-router/src/main/kotlin/tokens/AuthenticatedClaims.kt:84]: Fixed

File: prime-router/src/main/kotlin/tokens/Scope.kt (hunk 16 above)
Comment (jimduff-usds) [reply prime-router/src/main/kotlin/tokens/Scope.kt:27]: I'll leave it here because this is a scope and this is the Scope class.

File: prime-router/src/main/kotlin/tokens/OktaAuthentication.kt (hunk 17 above)
Comment (jimduff-usds) [reply prime-router/src/main/kotlin/tokens/OktaAuthentication.kt:139]: Yes, that's the ticket to move settings, lookuptable and other azure functions over to the new way.

File: prime-router/src/main/kotlin/tokens/Scope.kt (hunk 18 above)
Comment (jimduff-usds) [reply prime-router/src/main/kotlin/tokens/Scope.kt:58]: No, third position is currently defined by an enum set of limited values.   I thought about having *.*.* be the primeadmim capability, but decided it would be better to have the word 'primeadmin' specifically in the string, for clarity.

File: prime-router/src/main/kotlin/tokens/Scope.kt (hunk 19 above)
Comment (jimduff-usds) [reply prime-router/src/main/kotlin/tokens/Scope.kt:128]: done

File: prime-router/src/main/kotlin/tokens/Scope.kt (hunk 20 above)
Comment (jimduff-usds) [reply prime-router/src/main/kotlin/tokens/Scope.kt:136]: done
